components/places/android/src/main/java/org/mozilla/places/Bookmarks.kt

```diff
@@ -0,0 +1,407 @@
+}
+
+/**
+ * The methods provided by a read-only or a read-write bookmarks connection.
+ */
+interface BookmarkReadInterface {
+    /**
+     * Returns the bookmark subtree rooted at `rootGUID`. This differs from
+     * `getBookmark` in that it populates folder children.
+     *
+     * Specifically, any [BookmarkFolder]s in the returned value will have their
+     * `children` list populated, and not just `childGUIDs` (Note: if
+     * `recursive = false` is passed, then this is only performed for direct
+     * children, and not for grandchildren).
+     *
+     * @param rootGUID the GUID where to start the tree. Defaults to `BookmarkRoot.Root`'s id,
+     *                 e.g. fetching the entire bookmarks tree.
+     *
+     * @param recursive Whether or not to return more than a single
+     *                  level of children for folders. If false, then
+     *                  any folders which are children of the requested
+     *                  node will *only* have their `childGUIDs`
+     *                  populated, and *not* their `children`. Defaults to
+     *                  true.
+     *
+     * @return The bookmarks tree starting at `rootGUID`, or null if the provided
+     *         id didn't refer to a known bookmark item.
+     */
+    fun getBookmarksTree(rootGUID: String = BookmarkRoot.Root.id, recursive: Boolean = true): BookmarkTreeNode?
+
+    /**
+     * Returns the information about the bookmark with the provided id. This differs from
+     * `getBookmarksTree` in that it does not populate the `children` list if `guid` refers
+     * to a folder (only it's `childGUIDs` list).
+     *
+     * @param guid the guid of the bookmark to fetch.
+     * @return The bookmark node, or null if the provided
+     *         guid didn't refer to a known bookmark item.
+     */
+    fun getBookmark(guid: String): BookmarkTreeNode?
+
+    /**
+     * Returns the list of bookmarks with the provided URL.
+     *
+     * Note that if the URL is not percent-encoded/punycoded, that will be performed
+     * internally, and so the returned bookmarks may not have an identical URL to
+     * the one passed in (however, it will be the same according to the
+     * [URL standard](https://url.spec.whatwg.org/)).
+     *
+     * @param url The url to search for.
+     * @return A list of bookmarks that have the requested URL.
+     */
+    fun getBookmarksWithUrl(url: String): List<BookmarkTreeNode>
+}
+
+/**
+ * The methods provided by a bookmarks connection with write capabilities.
+ */
+interface BookmarkWriteInterface : BookmarkReadInterface {
+
+    /**
+     * Delete the bookmark with the provided GUID.
+     *
+     * If the requested bookmark is a folder, all children of
+     * bookmark are deleted as well, recursively.
+     *
+     * @param guid The GUID of the bookmark to delete
+     * @return Whether or not the bookmark existed.
+     */
+    fun deleteBookmark(guid: String): Boolean
```

Comment:
I'd prefer to see two different methods here: `deleteFolder` and `deleteBookmark`, and have them check that the passed in guid is indeed referring to a folder/bookmark. Accidentally calling `deleteBookmark` on a folder when you didn't intend to, and having all of its children deleted (and those deletions synced) could be a very costly (and possibly hard to find) bug.

Reply:
Hmm, I'm unsure how this issue would happen in practice, to be honest....

Reply:
I think I'll allow the caller to optionally specify what `type` they expect the bookmark to be, but that's mainly because it's the path of least resistance for me and not because I actually really think it will be used.

Reply:
Eh, as discussed in slack I ended up deciding against this since I think it would cause bad behavior.